lib/interface_quda.cpp

```diff
@@ -3252,346 +3230,389 @@ void invertQuda(void *hp_x, void *hp_b, QudaInvertParam *param)
-      for(int i=0; i < param->num_src; i++) {
-        cudaColorSpinorField tmp((in->Component(i)));
-        dirac.Mdag(in->Component(i), tmp);
-      }
-    } else if (!mat_solution && direct_solve) { // perform the first of two solves: A^dag y = b
-      DiracMdag m(dirac), mSloppy(diracSloppy), mPre(diracPre);
-      SolverParam solverParam(*param);
-      Solver *solve = Solver::create(solverParam, m, mSloppy, mPre, mPre, profileInvert);
-      solve->blocksolve(*out,*in);
-      for(int i=0; i < param->num_src; i++) {
-        blas::copy(in->Component(i), out->Component(i));
-      }
-      delete solve;
-      solverParam.updateInvertParam(*param);
+    for (int d = 0; d < CommKey::n_dim; d++) {
+      gauge_param->X[d] /= split_key[d];
+      gauge_param->ga_pad /= split_key[d];
     }
 
-    if (direct_solve) {
-      DiracM m(dirac), mSloppy(diracSloppy), mPre(diracPre);
-      SolverParam solverParam(*param);
-      Solver *solve = Solver::create(solverParam, m, mSloppy, mPre, mPre, profileInvert);
-      solve->blocksolve(*out,*in);
-      delete solve;
-      solverParam.updateInvertParam(*param);
-    } else if (!norm_error_solve) {
-      DiracMdagM m(dirac), mSloppy(diracSloppy), mPre(diracPre);
-      SolverParam solverParam(*param);
-      Solver *solve = Solver::create(solverParam, m, mSloppy, mPre, mPre, profileInvert);
-      solve->blocksolve(*out,*in);
-      delete solve;
-      solverParam.updateInvertParam(*param);
-    } else { // norm_error_solve
-      DiracMMdag m(dirac), mSloppy(diracSloppy), mPre(diracPre);
-      errorQuda("norm_error_solve not supported in multi source solve");
-      // cudaColorSpinorField tmp(*out);
-      // SolverParam solverParam(*param);
-      // Solver *solve = Solver::create(solverParam, m, mSloppy, mPre, profileInvert);
-      //(*solve)(tmp, *in); // y = (M M^\dag) b
-      // dirac.Mdag(*out, tmp);  // x = M^dag y
-      // delete solve;
-      // solverParam.updateInvertParam(*param,i,i);
+    for (int n = 0; n < param->num_src_per_sub_partition; n++) {
+      auto first = _h_x.begin() + n * num_sub_partition;
+      auto last = _h_x.begin() + (n + 1) * num_sub_partition;
+      std::vector<ColorSpinorField *> _v_x(first, last);
+      join_field(_v_x, *_collect_x[n], split_key, pc_type);
     }
 
-    if (getVerbosity() >= QUDA_VERBOSE){
-      for(int i=0; i < param->num_src; i++) {
-        double nx = blas::norm2(x->Component(i));
-        printfQuda("Solution %i = %g\n",i, nx);
-      }
+    for (auto p : _collect_b) { delete p; }
+    for (auto p : _collect_x) { delete p; }
+
+    for (auto p : _h_x) { delete p; }
+    for (auto p : _h_b) { delete p; }
+
+    if (!is_staggered) {
+      delete in;
+      delete collected_gauge;
+    } else {
+      delete milc_fatlink_field;
+      delete milc_longlink_field;
+      delete collected_milc_fatlink_field;
+      delete collected_milc_longlink_field;
     }
 
+    if (input_clover) { delete input_clover; }
```

Comment:
need to put my usual: This should be smart pointers comment here.

Reply:
I would prefer to put this as clean up for later?